File: Binary Search/410. Split Array Largest Sum/Solution.py
from typing import List

class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        # N : len(nums)
        # Time Complexity: O(N)
        # Space Complexity: O(1)
        res = float('inf')
        l, r = max(nums), sum(nums)
        while l <= r:
            mid = (r + l) // 2
            summ = 0
            subCount = 1
            for num in nums:
                summ += num
                if summ > mid:
                    summ = num
                    subCount += 1
            if subCount <= k:
                r = mid - 1
                res = mid
            else:
                l = mid + 1
        return res

File: Binary Search/410. Split Array Largest Sum/Solution.java
class Solution {
    public int splitArray(int[] nums, int k) {
        // N : len(nums)
        // Time Complexity: O(N)
        // Space Complexity: O(1)
        int res = Integer.MAX_VALUE;
        int l = Arrays.stream(nums).max().getAsInt();
        int r = Arrays.stream(nums).sum();
        while (l <= r) {
            int mid = (r + l) / 2;
            int subCount = 1;
            int summ = 0;
            for (int num: nums) {
                summ += num;
                if (summ > mid) {
                    summ = num;
                    subCount++;
                }
            }
            if (subCount <= k) {
                res = mid;
                r = mid - 1;
            } else l = mid + 1;
        }
        return res;
    }
}